File: app/src/main/java/fhaachen/snakegame/ui/GameStage.java
package fhaachen.snakegame.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.Rect;
import android.graphics.RectF;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.os.Handler;
import android.os.Looper;
import android.view.Display;
import android.view.MotionEvent;
import android.view.Surface;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import java.util.Random;
import java.util.stream.IntStream;

import fhaachen.snakegame.R;
import fhaachen.snakegame.model.Controls;
import fhaachen.snakegame.model.Snake;
import fhaachen.snakegame.model.Theme;

@SuppressLint("ViewConstructor")
public class GameStage extends SurfaceView implements Runnable, DialogInterface.OnDismissListener {
    private int numBlocksWide = 40;
    private long fps = 7;
    private final SurfaceHolder surfaceHolder;
    private final Paint paint;
    private Thread thread = null;
    private volatile boolean isRunning;
    private volatile boolean isPlaying;
    private boolean pauseMenuShown;
    private final int screenX;
    private final int screenY;
    private final int snakeBlockSize;
    private final int numBlocksHigh;
    private long nextFrameTime;
    private final int maxBlocksOnScreen;
    private final Display display;
    private final Controls.Mode controlMode;
    private final AppCompatActivity activity;

    private Snake snake;
    private final Controls controls;
    private final Rect food;
    private int score;

    //Resource strings
    private final String menuTitle;
    private final String currentScoreMsg;

    //Colors
    private int scoreTextColor;
    private final int snakeColor;
    private final int foodColor;
    private final int controllersColor;

    //Bitmaps
    private Bitmap backgroundBitmap;
    private Bitmap foodBitmap;
    private Bitmap snakeHeadBitmap;
    private Bitmap snakeBodyBitmap;

    public GameStage(Context context) {
        super(context);
        //Activity
        activity = (AppCompatActivity) getContext();

        //Context variables for later use
        Resources contextResources = activity.getResources();
        Resources.Theme contextTheme = activity.getTheme();

        //Shared preferences
        SharedPreferences sharedPref = activity.getPreferences(Context.MODE_PRIVATE);

        //Set resource strings
        menuTitle = activity.getString(R.string.app_name);
        currentScoreMsg = activity.getString(R.string.current_score);
        String settingTheme = activity.getString(R.string.setting_theme);

        //Set colors
        snakeColor = contextResources.getColor(R.color.snake, contextTheme);
        foodColor = contextResources.getColor(R.color.food, contextTheme);
        controllersColor = contextResources.getColor(R.color.controllers, contextTheme);
        int textColorLight = contextResources.getColor(R.color.textColorLight, contextTheme);
        int textColorDark = contextResources.getColor(R.color.textColorDark, contextTheme);

        //Default settings
        String defaultTheme = contextResources.getString(R.string.setting_theme_default);
        String defaultControlMode = contextResources.getString(R.string.setting_control_default);

        //FIXME: Example for saving theme setting
        /*SharedPreferences.Editor editor = sharedPref.edit();
        editor.putString(activity.getString(R.string.setting_theme), "WATER");
        editor.apply();*/

        //Theme switch
        Theme theme = Theme.valueOf(sharedPref.getString(activity.getString(R.string.setting_theme), defaultTheme));
        switch (theme) {
            case GRASS:
                backgroundBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.background_grass);
                foodBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.food_apple);
                snakeHeadBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_head);
                snakeBodyBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_body);
                scoreTextColor = textColorDark;
                break;
            case WATER:
                backgroundBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.background_water_new);
                foodBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.food_fish);
                snakeHeadBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_water_head);
                snakeBodyBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_water_body);
                scoreTextColor = textColorLight;
                break;
        }

        //Orientation lock
        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LOCKED);

        // Get the pixel dimensions of the screen
        display = activity.getWindowManager().getDefaultDisplay();
        // Initialize the result into a Point object
        Point size = new Point();
        display.getSize(size);
        //Set screen size
        screenX = size.x;
        screenY = size.y;

        surfaceHolder = getHolder();
        paint = new Paint();

        //Resize when using portrait mode
        if (screenX < screenY) {
            numBlocksWide = 20;
        }
        snakeBlockSize = screenX / numBlocksWide;
        numBlocksHigh = screenY / snakeBlockSize;
        maxBlocksOnScreen = numBlocksWide * numBlocksHigh;

        //FIXME: Example for saving control setting
        /*SharedPreferences.Editor editor = sharedPref.edit();
        editor.putString(activity.getString(R.string.setting_control), "GESTURES");
        editor.apply();*/

        //Prepares the button draw if needed
        controlMode = Controls.Mode.valueOf(sharedPref.getString(activity.getString(R.string.setting_control), defaultControlMode));
        if (controlMode == Controls.Mode.BUTTONS) {
            int controlButtonSize = snakeBlockSize * 3;
            int controlsY = screenY - (controlButtonSize * 3) - snakeBlockSize;
            controls = new Controls(snakeBlockSize, controlsY, controlButtonSize);
        } else {
            controls = null;
        }

        food = new Rect();

        nextFrameTime = System.currentTimeMillis();
    }

    @Override
    public void run() {
        while (isRunning) {
            if (updateRequired()) {
                if (isPlaying) {
                    update();
                }
                draw();
            }
        }
    }

    /**
     * Pauses the game
     */
    public void pause() {
        isRunning = false;
        try {
            thread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    /**
     * Resumes the game after counter hits zero
     */
    public void resume() {
        isRunning = true;
        thread = new Thread(this);
        thread.start();
    }

    @SuppressLint("InflateParams")
    // Pass null as the parent view because its going in the dialog layout
    public void showPauseDialog() {
        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR);
        pauseMenuShown = true;
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);

        // Inflate and set the layout for the dialog
        View view = activity.getLayoutInflater().inflate(R.layout.pause_menu, null);
        builder.setView(view)
                .setTitle(menuTitle)
                .setPositiveButton(R.string.play, (dialog, id) -> startGameAndClosePauseMenu())
                .setNegativeButton(R.string.exit, (dialog, id) -> activity.finishAndRemoveTask())
                .setOnDismissListener(this);

        TextView lastScoreLabel = view.findViewById(R.id.your_score_label);
        TextView lastScore = view.findViewById(R.id.your_score);
        if (score != 0) {
            lastScoreLabel.setVisibility(VISIBLE);
            lastScore.setVisibility(VISIBLE);
            lastScore.setText(String.valueOf(score));
        } else {
            lastScoreLabel.setVisibility(INVISIBLE);
            lastScore.setVisibility(INVISIBLE);
        }

        runOnUiThread(() -> {
            builder.create();
            builder.show();
        });
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        startGameAndClosePauseMenu();
    }

    private static void runOnUiThread(Runnable runnable) {
        Handler UIHandler = new Handler(Looper.getMainLooper());
        UIHandler.post(runnable);
    }

    private void startGameAndClosePauseMenu() {
        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LOCKED);
        startGame();
        pauseMenuShown = false;
        isRunning = true;
    }

    /**
     * Sets all variables needed for game start
     */
    private void startGame() {
        if (!isPlaying) {
            snake = new Snake(
                    numBlocksWide / 2,
                    numBlocksHigh / 2,
                    maxBlocksOnScreen);

            spawnFood();
            score = 0;
            nextFrameTime = System.currentTimeMillis();
            isPlaying = true;
            fps = 7;
        }
    }

    /**
     * Spawns food at a random location but not inside the snake
     */
    private void spawnFood() {
        Random random = new Random();
        int randomX;
        int randomY;

        do {
            randomX = random.nextInt(numBlocksWide - 1) + 1;
            randomY = random.nextInt(numBlocksHigh - 1) + 1;
        } while (positionInsideSnake(randomX, randomY));

        int x = randomX * snakeBlockSize;
        int y = randomY * snakeBlockSize;
        food.set(
                x,
                y,
                x + snakeBlockSize,
                y + snakeBlockSize);
    }

    /**
     * Checks if the given coordinates are included in the snake
     *
     * @param x coordinate to be checked
     * @param y coordinate to be checked
     * @return true if the coordinate is included
     */
    private boolean positionInsideSnake(int x, int y) {
        return IntStream.of(snake.getBodyXs()).anyMatch(snakeX -> snakeX == x) && IntStream.of(snake.getBodyYs()).anyMatch(snakeY -> snakeY == y);
    }

    /**
     * Consumes food and spawns new
     */
    private void eatFood() {
        score++;
        if (score < (maxBlocksOnScreen - 1)) {
            spawnFood();
            snake.increaseSize();
            if (score != 0 && score % 4 == 0 && fps <= 20) {
                fps++;
            }
        } else {
            isPlaying = false;
        }
    }

    /**
     * Evaluates if a screen update is required
     *
     * @return true if update is required
     */
    private boolean updateRequired() {
        if (!pauseMenuShown && nextFrameTime <= System.currentTimeMillis()) {
            nextFrameTime = System.currentTimeMillis() + 1000 / fps;
            return true;
        }
        return false;
    }

    /**
     * Updates the game state
     */
    private void update() {
        if ((snake.getHeadX() * snakeBlockSize) == food.left && (snake.getHeadY() * snakeBlockSize) == food.top) {
            eatFood();
        }

        snake.moveSnake();

        if (detectDeath()) {
            isPlaying = false;
            snake = null;
        }
    }

    /**
     * Evaluates if the player is dead
     *
     * @return true if death is detected
     */
    private boolean detectDeath() {
        // Hit the screen edge
        if (snake.getHeadX() == -1 || snake.getHeadX() >= numBlocksWide + 1 || snake.getHeadY() == -1 || snake.getHeadY() == numBlocksHigh + 1) {
            return true;
        }

        // Hit itself
        for (int i = snake.getSnakeLength(); i > 0; i--) {
            if ((i > 4)
                    && (snake.getHeadX() == snake.getBodyX(i))
                    && (snake.getHeadY() == snake.getBodyY(i))) {
                return true;
            }
        }

        // Hit nothing
        return false;
    }

    /**
     * Draws the game field
     */
    private void draw() {
        if (surfaceHolder.getSurface().isValid()) {
            Canvas canvas = surfaceHolder.lockCanvas();
            // Set background image
            canvas.drawBitmap(backgroundBitmap, null, new RectF(0, 0, screenX, screenY), null);

            if (isPlaying && !pauseMenuShown) {
                drawGame(canvas, paint);
            } else {
                drawStart();
            }

            surfaceHolder.unlockCanvasAndPost(canvas);
        }
    }

    private void drawStart() {

        showPauseDialog();


    }

    private void drawGame(Canvas canvas, Paint paint) {
        // Set controls color
        paint.setColor(controllersColor);

        // Draw controls if needed
        if (controlMode == Controls.Mode.BUTTONS && controls != null) {
            for (Rect control : controls.getButtons()) {
                canvas.drawRect(
                        control.left,
                        control.top,
                        control.right,
                        control.bottom,
                        paint);
            }
        }

        // Draw food
        paint.setColor(foodColor);
        Rect foodRect = new Rect(food.left, food.top, food.right, food.bottom);
        canvas.drawBitmap(foodBitmap, null, foodRect, paint);

        // Set snake color
        paint.setColor(snakeColor);

        // Draw the snake
        for (int i = 0; i < snake.getSnakeLength() + 1; i++) {
            Rect snakeRect = new Rect(snake.getBodyX(i) * snakeBlockSize,
                    (snake.getBodyY(i) * snakeBlockSize),
                    (snake.getBodyX(i) * snakeBlockSize) + snakeBlockSize,
                    (snake.getBodyY(i) * snakeBlockSize) + snakeBlockSize);
            if (i == 0) {
                canvas.drawBitmap(snakeHeadBitmap, null, snakeRect, paint);
            } else {
                canvas.drawBitmap(snakeBodyBitmap, null, snakeRect, paint);
            }
        }

        // Scale the HUD text
        paint.setTextSize(70);
        paint.setColor(scoreTextColor);
        canvas.drawText(String.format(currentScoreMsg, score), 10, 60, paint);
    }

    /**
     * Method for button control
     *
     * @param motionEvent event from touch
     * @return result of super call
     */
    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent motionEvent) {
        if (motionEvent.getAction() == MotionEvent.ACTION_DOWN) {
            if (controlMode == Controls.Mode.BUTTONS && isPlaying) {
                int posX = Math.round(motionEvent.getX());
                int posY = Math.round(motionEvent.getY());

                if (controls.getButton(Controls.Button.LEFT).contains(posX, posY)) {
                    snake.setDirectionLeft();
                } else if (controls.getButton(Controls.Button.UP).contains(posX, posY)) {
                    snake.setDirectionUp();
                } else if (controls.getButton(Controls.Button.RIGHT).contains(posX, posY)) {
                    snake.setDirectionRight();
                } else if (controls.getButton(Controls.Button.DOWN).contains(posX, posY)) {
                    snake.setDirectionDown();
                }
            } else {
                startGame();
            }
        }
        return super.onTouchEvent(motionEvent);
    }

    /**
     * Method for tilt control
     *
     * @param event         event from the sensor
     * @param accelerometer Sensor that fired the event
     */
    public void onSensorChanged(SensorEvent event, Sensor accelerometer) {
        if (event.sensor == accelerometer) {
            if (controlMode == Controls.Mode.TILT && isPlaying) {
                int x = Math.round(event.values[0]);
                int y = Math.round(event.values[1]);
                boolean xStrongerThanY = Math.abs(x) > Math.abs(y);

                switch (display.getRotation()) {
                    case Surface.ROTATION_0:
                        //LEFT  : +x
                        //UP    : -y
                        //RIGHT : -x
                        //DOWN  : +y
                        if (xStrongerThanY) {
                            if (x > 0) {
                                snake.setDirectionLeft();
                            } else {
                                snake.setDirectionRight();
                            }
                        } else {
                            if (y > 0) {
                                snake.setDirectionDown();
                            } else {
                                snake.setDirectionUp();
                            }
                        }
                        break;
                    case Surface.ROTATION_90:
                        //LEFT  : -y
                        //UP    : -x
                        //RIGHT : +y
                        //DOWN  : +x
                        if (xStrongerThanY) {
                            if (x > 0) {
                                snake.setDirectionDown();
                            } else {
                                snake.setDirectionUp();
                            }
                        } else {
                            if (y > 0) {
                                snake.setDirectionRight();
                            } else {
                                snake.setDirectionLeft();
                            }
                        }
                        break;
                    case Surface.ROTATION_180:
                        //LEFT  : -x
                        //UP    : +y
                        //RIGHT : +x
                        //DOWN  : -y
                        if (xStrongerThanY) {
                            if (x > 0) {
                                snake.setDirectionRight();
                            } else {
                                snake.setDirectionLeft();
                            }
                        } else {
                            if (y > 0) {
                                snake.setDirectionUp();
                            } else {
                                snake.setDirectionDown();
                            }
                        }
                        break;
                    case Surface.ROTATION_270:
                        //LEFT  : +y
                        //UP    : +x
                        //RIGHT : -y
                        //DOWN  : -x
                        if (xStrongerThanY) {
                            if (x > 0) {
                                snake.setDirectionUp();
                            } else {
                                snake.setDirectionDown();
                            }
                        } else {
                            if (y > 0) {
                                snake.setDirectionLeft();
                            } else {
                                snake.setDirectionRight();
                            }
                        }
                        break;
                }
            }
        }
    }

    /**
     * Method for fling/gesture control
     *
     * @param velocityX range swiped on x axis
     * @param velocityY range swiped on y axis
     * @return true when event is consumed
     */
    public boolean onFling(float velocityX, float velocityY) {
        if (controlMode == Controls.Mode.GESTURES) {
            if (Math.abs(velocityX) > Math.abs(velocityY)) {
                if (velocityX > 0) {
                    snake.setDirectionRight();
                } else {
                    snake.setDirectionLeft();
                }
            } else {
                if (velocityY > 0) {
                    snake.setDirectionDown();
                } else {
                    snake.setDirectionUp();
                }
            }
        } else {
            return false;
        }
        return true;
    }
}

File: app/src/main/java/fhaachen/snakegame/ui/GameStage.kt
package fhaachen.snakegame.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.ActivityInfo
import android.graphics.*
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.os.Handler
import android.os.Looper
import android.view.*
import android.widget.RadioButton
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ContextThemeWrapper
import androidx.core.content.ContextCompat.startActivity
import fhaachen.snakegame.R
import fhaachen.snakegame.enums.ControlMode
import fhaachen.snakegame.enums.Theme
import fhaachen.snakegame.helper.ControlsHelper
import fhaachen.snakegame.helper.ScoreHelper.getHighScore
import fhaachen.snakegame.helper.ScoreHelper.getLastScore
import fhaachen.snakegame.helper.ScoreHelper.saveScore
import fhaachen.snakegame.helper.SharedPreferencesHelper.getSharedPreference
import fhaachen.snakegame.helper.SharedPreferencesHelper.setSharedPreference
import fhaachen.snakegame.helper.SnakeHelper.detectDeath
import fhaachen.snakegame.model.Controls
import fhaachen.snakegame.model.Snake
import java.util.*
import java.util.stream.IntStream

class GameStage
constructor(context: Context?) : SurfaceView(context), Runnable, DialogInterface.OnDismissListener {
    //Misc
    private val surfaceHolder: SurfaceHolder
    private val paint: Paint = Paint()
    private var thread: Thread? = null
    private val activity: AppCompatActivity = getContext() as AppCompatActivity
    private val applicationContext: Context = getContext().applicationContext

    //Sizing
    private var numBlocksWide = 40
    private val screenX: Int
    private val screenY: Int
    private val snakeBlockSize: Int
    private val numBlocksHigh: Int
    private val maxBlocksOnScreen: Int
    private val screen: Display = activity.windowManager.defaultDisplay

    //Timing
    private var fps: Long = 7
    private var nextFrameTime: Long

    //Game states
    @Volatile
    private var isRunning = false

    @Volatile
    private var isPlaying = false
    private var pauseMenuShown = false

    //Game elements
    private var snake: Snake? = null
    private var controls: Controls? = null
    private lateinit var controlMode: ControlMode
    private var food: Rect = Rect()
    private var score = 0

    //Colors
    private var scoreTextColor = 0

    //Bitmaps
    private lateinit var backgroundBitmap: Bitmap
    private lateinit var foodBitmap: Bitmap
    private lateinit var snakeHeadBitmap: Bitmap
    private lateinit var snakeBodyBitmap: Bitmap

    override fun run() {
        while (isRunning) {
            if (updateRequired()) {
                if (isPlaying) {
                    update()
                }
                draw()
            }
        }
    }

    /**
     * Pauses the game
     */
    fun pause() {
        isRunning = false
        try {
            thread!!.join()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    /**
     * Resumes the game after counter hits zero
     */
    fun resume() {
        isRunning = true
        thread = Thread(this)
        thread!!.start()
    }

    @SuppressLint("InflateParams")
    fun showMenuDialog() {
        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR
        pauseMenuShown = true
        val builder = AlertDialog.Builder(ContextThemeWrapper(context, R.style.ScoreTheme))
        // Inflate and set the layout for the dialog
        val view = activity.layoutInflater.inflate(R.layout.pause_menu, null)
        builder.setView(view)
                .setTitle(applicationContext.getString(R.string.app_name))
                .setPositiveButton(R.string.button_play) { _: DialogInterface?, _: Int -> startGameAndClosePauseMenu() }
                .setNegativeButton(R.string.button_exit) { _: DialogInterface?, _: Int -> activity.finishAndRemoveTask() }
                .setOnDismissListener(this)
        //Scores
        val lastScore = view.findViewById<TextView>(R.id.your_score)
        val highScore = view.findViewById<TextView>(R.id.highscore)

        highScore.setOnClickListener {
            composeMessage(applicationContext.getString(R.string.share_highscore, getSharedPreference(context, R.string.save_highscore, 0)))
        }

        lastScore.text = getLastScore(context)
        highScore.text = getHighScore(context)
        //Set theme radioButton
        val themeRadioButton: RadioButton
        val theme = Theme.valueOf(getSharedPreference(context, R.string.setting_theme, Theme.GRASS.toString()))
        themeRadioButton = when (theme) {
            Theme.WATER -> view.findViewById(R.id.theme_water_button)
            else -> view.findViewById(R.id.theme_grass_button)
        }
        runOnUiThread(Runnable { themeRadioButton.isChecked = true })
        val controlModeRadioButton: RadioButton
        val controlMode = ControlMode.valueOf(getSharedPreference(context, R.string.setting_control, ControlMode.GESTURES.toString()))
        controlModeRadioButton = when (controlMode) {
            ControlMode.BUTTONS -> view.findViewById(R.id.control_buttons_button)
            ControlMode.TILT -> view.findViewById(R.id.control_tilt_button)
            else -> view.findViewById(R.id.control_swype_button)
        }
        runOnUiThread(Runnable { controlModeRadioButton.isChecked = true })
        runOnUiThread(Runnable {
            builder.create()
            builder.show()
        })
    }

    /**
     * Share message
     */
    private fun composeMessage(message: String) {
        val intent = Intent(Intent.ACTION_SEND).apply {
            putExtra(Intent.EXTRA_TEXT, message)
            type = "text/plain"
        }
        startActivity(context, Intent.createChooser(intent, "Share"), null)
    }

    override fun onDismiss(dialog: DialogInterface) {
        startGameAndClosePauseMenu()
    }

    private fun startGameAndClosePauseMenu() {
        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LOCKED
        startGame()
        pauseMenuShown = false
        isRunning = true
    }

    /**
     * Sets all variables needed for game start
     */
    private fun startGame() {
        if (!isPlaying) {
            snake = Snake(
                    numBlocksWide / 2,
                    numBlocksHigh / 2,
                    maxBlocksOnScreen)
            spawnFood()
            score = 0
            nextFrameTime = System.currentTimeMillis()
            isPlaying = true
            fps = 7
        }
    }

    /**
     * Spawns food at a random location but not inside the snake
     */
    private fun spawnFood() {
        val random = Random()
        var randomX: Int
        var randomY: Int
        do {
            randomX = random.nextInt(numBlocksWide - 1) + 1
            randomY = random.nextInt(numBlocksHigh - 1) + 1
        } while (positionInsideSnake(randomX, randomY))
        val x = randomX * snakeBlockSize
        val y = randomY * snakeBlockSize
        food[x, y, x + snakeBlockSize] = y + snakeBlockSize
    }

    /**
     * Checks if the given coordinates are included in the snake
     *
     * @param x coordinate to be checked
     * @param y coordinate to be checked
     * @return true if the coordinate is included
     */
    private fun positionInsideSnake(x: Int, y: Int): Boolean {
        return IntStream.of(*snake!!.bodyXs).anyMatch { snakeX: Int -> snakeX == x } && IntStream.of(*snake!!.bodyYs).anyMatch { snakeY: Int -> snakeY == y }
    }

    /**
     * Consumes food and spawns new
     */
    private fun eatFood() {
        score++
        if (score < maxBlocksOnScreen - 1) {
            spawnFood()
            snake!!.increaseSize()
            if (score != 0 && score % 4 == 0 && fps <= 20) {
                fps++
            }
        } else {
            // Screen is full with the snake
            isPlaying = false
        }
    }

    /**
     * Evaluates if a screen update is required
     *
     * @return true if update is required
     */
    private fun updateRequired(): Boolean {
        if (!pauseMenuShown && nextFrameTime <= System.currentTimeMillis()) {
            nextFrameTime = System.currentTimeMillis() + 1000 / fps
            return true
        }
        return false
    }

    /**
     * Updates the game state
     */
    private fun update() {
        if (snake!!.headX * snakeBlockSize == food.left && snake!!.headY * snakeBlockSize == food.top) {
            eatFood()
        }
        snake!!.moveSnake()
        if (detectDeath(snake!!, numBlocksWide, numBlocksHigh)) {
            saveScore(context, score)
            isPlaying = false
            snake = null
        }
    }

    /**
     * Draws the game field
     */
    private fun draw() {
        if (surfaceHolder.surface.isValid) {
            val canvas = surfaceHolder.lockCanvas()
            drawBackgroundImage(canvas)
            // Draw game or menu?
            if (isPlaying && !pauseMenuShown) {
                drawGame(canvas)
            } else {
                showMenuDialog()
            }
            surfaceHolder.unlockCanvasAndPost(canvas)
        }
    }

    /**
     * Draws the background image
     */
    private fun drawBackgroundImage(canvas: Canvas) {
        // Set background image
        canvas.drawBitmap(backgroundBitmap, null, RectF(0F, 0F, screenX.toFloat(), screenY.toFloat()), null)
    }

    private fun drawGame(canvas: Canvas) {
        drawControls(canvas)

        drawFood(canvas)

        drawSnake(canvas)

        drawCurrentScore(canvas)
    }

    /**
     * Draws the Control buttons depending on current [controlMode]
     */
    private fun drawControls(canvas: Canvas) {
        // Set controls color
        paint.color = applicationContext.getColor(R.color.controllers)
        // Draw controls if needed
        if (controlMode === ControlMode.BUTTONS && controls != null) {
            for (control in controls!!.buttons) {
                canvas.drawRect(
                        control!!.left.toFloat(),
                        control.top.toFloat(),
                        control.right.toFloat(),
                        control.bottom.toFloat(),
                        paint)
            }
        }
    }

    /**
     * Draws current [food]
     */
    private fun drawFood(canvas: Canvas) {
        // Draw food
        paint.color = applicationContext.getColor(R.color.food)
        val foodRect = Rect(food.left, food.top, food.right, food.bottom)
        canvas.drawBitmap(foodBitmap, null, foodRect, paint)
    }

    /**
     * Draws current [snake]
     */
    private fun drawSnake(canvas: Canvas) {
        // Set snake color
        paint.color = applicationContext.getColor(R.color.snake)
        // Draw the snake
        for (i in 0 until snake!!.snakeLength + 1) {
            val snakeRect = Rect(snake!!.getBodyX(i) * snakeBlockSize,
                    snake!!.getBodyY(i) * snakeBlockSize,
                    snake!!.getBodyX(i) * snakeBlockSize + snakeBlockSize,
                    snake!!.getBodyY(i) * snakeBlockSize + snakeBlockSize)
            if (i == 0) {
                canvas.drawBitmap(snakeHeadBitmap, null, snakeRect, paint)
            } else {
                canvas.drawBitmap(snakeBodyBitmap, null, snakeRect, paint)
            }
        }
    }

    /**
     * Draws the Current score label
     */
    private fun drawCurrentScore(canvas: Canvas) {
        // Set text properties
        paint.textSize = 70f
        paint.color = scoreTextColor
        // Draw the text
        canvas.drawText(String.format(applicationContext.getString(R.string.label_current_score), score), 10f, 60f, paint)
    }

    /**
     * Updates the theme to the one stored in SavedPreferences
     */
    private fun updateTheme() {
        val applicationContext = context.applicationContext
        val contextResources = applicationContext.resources
        val contextTheme = applicationContext.theme
        val defaultTheme = contextResources.getString(R.string.setting_theme_default)
        when (Theme.valueOf(getSharedPreference(context, R.string.setting_theme, defaultTheme))) {
            Theme.GRASS -> {
                backgroundBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.background_grass)
                foodBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.food_apple)
                snakeHeadBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_head)
                snakeBodyBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_body)
                scoreTextColor = contextResources.getColor(R.color.textColorDark, contextTheme)
            }
            Theme.WATER -> {
                backgroundBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.background_water_new)
                foodBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.food_fish)
                snakeHeadBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_water_head)
                snakeBodyBitmap = BitmapFactory.decodeResource(contextResources, R.drawable.snake_water_body)
                scoreTextColor = contextResources.getColor(R.color.textColorLight, contextTheme)
            }
        }
    }

    /**
     * Updates the control mode to the one stored in SavedPreferences
     */
    private fun updateControlMode() {
        val applicationContext = context.applicationContext
        val contextResources = applicationContext.resources
        val defaultControlMode = contextResources.getString(R.string.setting_control_default)
        controlMode = ControlMode.valueOf(getSharedPreference(context, R.string.setting_control, defaultControlMode))
        controls = if (controlMode === ControlMode.BUTTONS) {
            val controlButtonSize = snakeBlockSize * 3
            val controlsY = screenY - controlButtonSize * 3 - snakeBlockSize
            Controls(snakeBlockSize, controlsY, controlButtonSize)
        } else {
            null
        }
    }

    /**
     * Method for button control
     *
     * @param motionEvent event from touch
     * @return result of super call
     */
    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(motionEvent: MotionEvent): Boolean {
        if (snake != null && ControlsHelper.onTouchEvent(motionEvent, controlMode, isPlaying, controls, snake!!)) {
            startGame()
        }
        return super.onTouchEvent(motionEvent)
    }

    /**
     * Method for tilt control
     *
     * @param event         event from the sensor
     * @param accelerometer Sensor that fired the event
     */
    fun onSensorChanged(event: SensorEvent, accelerometer: Sensor) {
        if (snake != null && isPlaying) {
            ControlsHelper.onSensorChanged(event, accelerometer, controlMode, screen.rotation, snake!!)
        }
    }

    /**
     * Method for fling/gesture control
     *
     * @param velocityX range swiped on x axis
     * @param velocityY range swiped on y axis
     * @return true when event is consumed
     */
    fun onFling(velocityX: Float, velocityY: Float): Boolean {
        if (snake != null) {
            return ControlsHelper.onFling(velocityX, velocityY, controlMode, isPlaying, snake!!)
        }
        return false
    }

    /**
     * Assigns its [Controls.ControlMode] to the pressed radio button and updates the controls if required
     *
     * @param v pressed View
     */
    fun onControlSelected(v: View) {
        when (v.id) {
            R.id.control_buttons_button -> updateControlModeIfRequired(ControlMode.BUTTONS.toString())
            R.id.control_swype_button -> updateControlModeIfRequired(ControlMode.GESTURES.toString())
            R.id.control_tilt_button -> updateControlModeIfRequired(ControlMode.TILT.toString())
        }
    }

    /**
     * Checks if the setting is different from the previous one and triggers an update accordingly
     *
     * @param value of the [Controls.ControlMode] to be used
     */
    private fun updateControlModeIfRequired(value: String) {
        val defaultControl = context.applicationContext.resources.getString(R.string.setting_control_default)
        if (getSharedPreference(context, R.string.setting_control, defaultControl) != value) {
            setSharedPreference(context, R.string.setting_control, value)
            updateControlMode()
        }
    }

    /**
     * Assigns its [Theme] to the pressed radio button and updates the theme if required
     *
     * @param v pressed View
     */
    fun onThemeSelected(v: View) {
        when (v.id) {
            R.id.theme_grass_button -> updateThemeIfRequired(Theme.GRASS.toString())
            R.id.theme_water_button -> updateThemeIfRequired(Theme.WATER.toString())
        }
    }

    /**
     * Checks if the setting is different from the previous one and triggers an update accordingly
     *
     * @param value of the [Theme] to be used
     */
    private fun updateThemeIfRequired(value: String) {
        val defaultTheme = context.applicationContext.resources.getString(R.string.setting_theme_default)
        if (getSharedPreference(context, R.string.setting_theme, defaultTheme) != value) {
            setSharedPreference(context, R.string.setting_theme, value)
            updateTheme()
        }
    }

    companion object {
        private fun runOnUiThread(runnable: Runnable) {
            val uiHandler = Handler(Looper.getMainLooper())
            uiHandler.post(runnable)
        }
    }

    init {
        //Theme switch
        updateTheme()

        //Orientation lock
        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LOCKED

        // Set screen size
        val size = Point()
        screen.getSize(size)
        screenX = size.x
        screenY = size.y

        surfaceHolder = holder

        if (screenX < screenY) {
            //Resize when using portrait mode
            numBlocksWide = 20
        }
        snakeBlockSize = screenX / numBlocksWide
        numBlocksHigh = screenY / snakeBlockSize
        maxBlocksOnScreen = numBlocksWide * numBlocksHigh
        //Prepares the button draw if needed
        updateControlMode()

        nextFrameTime = System.currentTimeMillis()
    }
}